flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoRequestCoordinator.java

```diff
@@ -0,0 +1,188 @@
+import org.apache.flink.runtime.webmonitor.stats.TaskStatsRequestCoordinator;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** A coordinator for triggering and collecting thread info stats of running operator subtasks. */
+public class ThreadInfoRequestCoordinator
+        extends TaskStatsRequestCoordinator<List<ThreadInfoSample>, OperatorThreadInfoStats> {
+
+    /**
+     * Creates a new coordinator for the job.
+     *
+     * @param executor Used to execute the futures.
+     * @param requestTimeout Time out after the expected sampling duration. This is added to the
+     *     expected duration of a request, which is determined by the number of samples and the
+     *     delay between each sample.
+     */
+    public ThreadInfoRequestCoordinator(Executor executor, long requestTimeout) {
+        super(executor, requestTimeout);
+    }
+
+    /**
+     * Triggers collection of thread info stats of an operator by combining thread info responses
+     * from given subtasks. A thread info response of a subtask in turn consists of {@code
+     * numSamples}, collected with {@code delayBetweenSamples} milliseconds delay between them.
+     *
+     * @param subtasksWithGateways Execution vertices together with TaskExecutors running them.
+     * @param numSamples Number of thread info samples to collect from each subtask.
+     * @param delayBetweenSamples Delay between consecutive samples (ms).
+     * @param maxStackTraceDepth Maximum depth of the stack traces collected within thread info
+     *     samples.
+     * @return A future of the completed thread info stats.
+     */
+    public CompletableFuture<OperatorThreadInfoStats> triggerThreadInfoRequest(
+            List<Tuple2<AccessExecutionVertex, CompletableFuture<TaskExecutorGateway>>>
+                    subtasksWithGateways,
+            int numSamples,
+            Time delayBetweenSamples,
+            int maxStackTraceDepth) {
+
+        checkNotNull(subtasksWithGateways, "Tasks to sample");
+        checkArgument(subtasksWithGateways.size() > 0, "No tasks to sample");
+        checkArgument(numSamples >= 1, "No number of samples");
+        checkArgument(maxStackTraceDepth >= 0, "Negative maximum stack trace depth");
+
+        // Execution IDs of running tasks
+        List<ExecutionAttemptID> runningSubtasksIds = new ArrayList<>();
+
+        // Check that all tasks are RUNNING before triggering anything. The
+        // triggering can still fail.
+        for (Tuple2<AccessExecutionVertex, CompletableFuture<TaskExecutorGateway>>
+                executionsWithGateway : subtasksWithGateways) {
+            AccessExecution execution = executionsWithGateway.f0.getCurrentExecutionAttempt();
```

Comment:
IF all we need are the attemptIds then we should only require that to be passed in.

Reply:
The motivation was to pin the task which is failing in case of errors:
https://github.com/apache/flink/pull/15054/files/de5180a21e4f51384895bee7e2527677aa8092c1#diff-803556ada30870d9cec571303d2d7d89a63f9980c26a79020a15ab239e2a837eR95
So it is somewhat related to the question raised in the previous comment about error handling.

Reply:
Filtering state before passing the `executionsWithGateway` now - that allows to only pass `AttemptID`.